src/lib/crypto/kimchi_bindings/stubs/src/pasta_fp_plonk_proof.rs

```diff
@@ -247,58 +251,89 @@ pub fn caml_pasta_fp_plonk_proof_example_with_foreign_field_mul(
+    // Constant Double Generic gate for result and quotient bounds
+    external_checks.add_high_bound_computation(&left_limbs[2]);
+    external_checks.add_high_bound_computation(&right_limbs[2]);
+    CircuitGate::extend_high_bounds(&mut gates, &mut next_row, &foreign_field_modulus);
+    gates.connect_cell_pair((15, 0), (0, 2)); // left2
+    gates.connect_cell_pair((15, 3), (0, 5)); // right2
+    external_checks.extend_witness_high_bounds_computation(&mut witness, &foreign_field_modulus);
+
+    // Left input multi-range-check
+    external_checks.add_multi_range_check(&left_limbs);
     CircuitGate::extend_multi_range_check(&mut gates, &mut next_row);
-    gates.connect_cell_pair((0, 6), (12, 0)); // carry1_lo
-    gates.connect_cell_pair((1, 5), (13, 0)); // product1_lo
-    gates.connect_cell_pair((1, 6), (14, 0)); // product1_hi_0
+    gates.connect_cell_pair((0, 0), (16, 0)); // left_input0
+    gates.connect_cell_pair((0, 1), (17, 0)); // left_input1
+    gates.connect_cell_pair((0, 2), (18, 0)); // left_input2
                                               // Witness updated below
 
-    // Result/remainder bound multi-range-check
+    // Right input multi-range-check
+    external_checks.add_multi_range_check(&right_limbs);
     CircuitGate::extend_multi_range_check(&mut gates, &mut next_row);
-    gates.connect_cell_pair((3, 0), (16, 0));
-    gates.connect_cell_pair((3, 1), (17, 0));
-    gates.connect_cell_pair((3, 2), (18, 0));
-    // Witness updated below
+    gates.connect_cell_pair((0, 3), (20, 0)); // right_input0
+    gates.connect_cell_pair((0, 4), (21, 0)); // right_input1
+    gates.connect_cell_pair((0, 5), (22, 0)); // right_input2
+                                              // Witness updated below
 
-    // Add witness for external multi-range checks (product1_lo, product1_hi_0, carry1_lo and result)
+    // Add witness for external multi-range checks:
+    // left and right limbs
     external_checks.extend_witness_multi_range_checks(&mut witness);
 
-    // Quotient bound multi-range-check
-    CircuitGate::extend_compact_multi_range_check(&mut gates, &mut next_row);
-    gates.connect_cell_pair((1, 3), (22, 1));
-    gates.connect_cell_pair((1, 4), (20, 0));
-    external_checks.extend_witness_compact_multi_range_checks(&mut witness);
+    // [result_bound, 0, 0]
+    // Bounds for result limb range checks
+    CircuitGate::extend_multi_range_check(&mut gates, &mut next_row);
+    gates.connect_cell_pair((6, 2), (24, 0)); // result_bound
+                                              // Witness updated below
+
+    // Multi-range check bounds for left and right inputs
+    let left_hi_bound =
+        foreign_field_mul::witness::compute_high_bound(&left_input, &foreign_field_modulus);
+    let right_hi_bound =
+        foreign_field_mul::witness::compute_high_bound(&right_input, &foreign_field_modulus);
+    external_checks.add_limb_check(&left_hi_bound.into());
```

Comment:
Where is `add_limb_check` happening for `remainder2` bound?

Reply:
Happens inside the gadget itself, as it is not the circuit writer's duty to decide when to call it, but it should happen inside each ffmul call
https://github.com/o1-labs/proof-systems/blob/4a0528472d20346441e664481718c7e27205be17/kimchi/src/circuits/polynomials/foreign_field_mul/witness.rs#L190